packages/grafana-ui/src/components/Menu/Menu.tsx

```diff
@@ -1,23 +1,97 @@
-import React from 'react';
+import React, { useEffect, useRef, useState } from 'react';
 import { css } from '@emotion/css';
 import { GrafanaTheme2 } from '@grafana/data';
 import { useStyles2 } from '../../themes';
+import { useEffectOnce } from 'react-use';
 
 /** @internal */
 export interface MenuProps extends React.HTMLAttributes<HTMLDivElement> {
   /** React element rendered at the top of the menu */
   header?: React.ReactNode;
   children: React.ReactNode;
   ariaLabel?: string;
+  onOpen?: (focusOnItem: (itemId: number) => void) => void;
+  onKeyDown?: React.KeyboardEventHandler;
 }
 
+const modulo = (a: number, n: number) => ((a % n) + n) % n;
+const UNFOCUSED = -1;
+type MenuItemElement = HTMLAnchorElement & HTMLButtonElement;
+
 /** @internal */
 export const Menu = React.forwardRef<HTMLDivElement, MenuProps>(
-  ({ header, children, ariaLabel, ...otherProps }, ref) => {
+  ({ header, children, ariaLabel, onOpen, onKeyDown, ...otherProps }, forwardedRef) => {
     const styles = useStyles2(getStyles);
 
+    const [focusedItem, setFocusedItem] = useState(UNFOCUSED);
+
+    const localRef = useRef<HTMLDivElement>(null);
+    const wrapperRef = (forwardedRef as React.RefObject<HTMLDivElement>) ?? localRef;
+
+    useEffect(() => {
+      const menuItems = wrapperRef?.current?.querySelectorAll(`[data-role="menuitem"]`);
+      (menuItems?.[focusedItem] as MenuItemElement)?.focus();
+      menuItems?.forEach((menuItem, i) => {
+        (menuItem as MenuItemElement).tabIndex = i === focusedItem ? 0 : -1;
+      });
+    }, [wrapperRef, focusedItem]);
+
+    useEffectOnce(() => {
+      const firstMenuItem = wrapperRef?.current?.querySelector(`[data-role="menuitem"]`) as MenuItemElement | null;
+      if (firstMenuItem) {
+        firstMenuItem.tabIndex = 0;
+      }
+      onOpen?.(setFocusedItem);
+    });
+
+    const handleKeys = (event: React.KeyboardEvent) => {
+      const menuItemsCount = wrapperRef?.current?.querySelectorAll('[data-role="menuitem"]').length ?? 0;
+
+      switch (event.key) {
+        case 'ArrowUp':
+          event.preventDefault();
+          event.stopPropagation();
```

Comment:
These can be called before the switch, so they're not repeated for each case.

Reply:
Would this not eat certain key-combinations that we want to propagate (e.g. global shortcuts)?

Reply:
True, didn't think about that, good point.